htmresearch/algorithms/sparse_net.py

```diff
@@ -0,0 +1,352 @@
+  """
+
+  def __init__(self,
+               inputDim=64,
+               outputDim=64,
+               batchSize=100,
+               numIterations=1000,
+               numLcaIterations=75,
+               learningRate=2.0,
+               decayCycle=100,
+               learningRateDecay=1.0,
+               lcaLearningRate=0.1,
+               thresholdDecay=0.95,
+               minThreshold=0.1,
+               thresholdType='soft',
+               verbosity=0,
+               showEvery=500,
+               seed=42):
+    """
+    Initializes the SparseNet.
+    :param inputDim:                (int)   (Flattened) dimension of input data
+    :param outputDim:               (int)   Output dimension
+    :param batchSize:               (int)   Batch size for training
+    :param numIterations:           (int)   Number of training iterations
+    :param numLcaIterations:        (int)   Number of iterations in LCA
+    :param learningRate:            (float) Learning rate
+    :param decayCycle:              (int)   Number of iterations between decays
+    :param learningRateDecay        (float) Learning rate decay rate
+    :param lcaLearningRate          (float) Learning rate in LCA
+    :param minThreshold:            (float) Minimum activation threshold
+                                            during decay
+    :param verbosity:               (int)   Verbosity level
+    :param seed:                    (int)   Seed for random number generators
+    """
+    self.inputDim = inputDim
+    self.outputDim = outputDim
+    self.batchSize = batchSize
+    self._reset()
+
+    # training parameters
+    self.numIterations = numIterations
+    self.learningRate = learningRate
+    self.decayCycle = decayCycle
+    self.learningRateDecay = learningRateDecay
+
+    # LCA parameters
+    self.numLcaIterations = numLcaIterations
+    self.lcaLearningRate = lcaLearningRate
+    self.thresholdDecay = thresholdDecay
+    self.minThreshold = minThreshold
+    self.thresholdType = thresholdType
+
+    # debugging
+    self.verbosity = verbosity
+    self.showEvery = showEvery
+    if seed is not None:
+      np.random.seed(seed)
+      random.seed(seed)
+
+
+  def train(self, inputData, reset=True):
+    """
+    Trains the SparseNet, with the provided data.
+
+    The reset parameter can be set to False if the network should not be
+    reset before training (for example for continuing a previous started
+    training).
+    :param inputData:   (array) Input data, of dimension (inputDim, numPoints)
+    :param reset:       (bool)  If set to false, keep basis and history
+    """
+    if not isinstance(inputData, np.ndarray):
+      inputData = np.array(inputData)
+
+    if reset:
+      self._reset()
+
+    for _ in xrange(self.numIterations):
+      self._iteration += 1
+
+      batch = self._getDataBatch(inputData)
+
+      # check input dimension, change if necessary
+      if batch.shape[0] != self.inputDim:
```

Comment:
Looks like inputDim can be modified here. Is it really a parameter the user set upfront then?